client/peerset/src/peer_store.rs

```diff
@@ -0,0 +1,140 @@
+/// the values above. In this setup the maximum allowed absolute value of `i32::MAX` becomes 0 in
+/// ~1100 seconds, and this is the maximun time records live in the hash map if they are not
+/// updated.
+const INVERSE_DECREMENT: i32 = 50;
+
+pub trait PeerReputationProvider {
+	/// Check whether the peer is banned.
+	fn is_banned(&self, peer_id: PeerId) -> bool;
+
+	/// Report the peer disconnect for reputation adjustement.
+	fn report_disconnect(&self, peer_id: PeerId, reason: DropReason);
+
+	/// Get the candidates for initiating outgoing connections.
+	fn outgoing_candidates(&self, count: usize, ignored: HashSet<&PeerId>) -> Vec<PeerId>;
+}
+
+pub struct PeerStoreHandle {}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+struct Reputation(i32);
+
+impl Default for Reputation {
+	fn default() -> Self {
+		Reputation(0)
+	}
+}
+
+impl Ord for Reputation {
+	// We define reverse order for reputation values.
+	fn cmp(&self, other: &Self) -> Ordering {
+		self.0.cmp(&other.0).reverse()
+	}
+}
+
+impl PartialOrd for Reputation {
+	fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+		Some(self.cmp(other))
+	}
+}
+
+impl Reputation {
+	fn is_banned(&self) -> bool {
+		self.0 < BANNED_THRESHOLD
+	}
+
+	fn add_reputation(&mut self, increment: i32) {
+		self.0 = self.0.saturating_add(increment);
+	}
+
+	fn decay(&mut self, seconds_passed: u32) {
+		let reputation = &mut self.0;
+
+		for _ in 0..seconds_passed {
+			let mut diff = *reputation / INVERSE_DECREMENT;
+			if diff == 0 && *reputation < 0 {
+				diff = -1;
+			} else if diff == 0 && *reputation > 0 {
+				diff = 1;
+			}
+
+			*reputation = reputation.saturating_sub(diff);
+
+			if *reputation == 0 {
+				break
+			}
+		}
+	}
+}
+
+struct PeerStoreInner {
+	reputations: HashMap<PeerId, Reputation>,
+}
+
+impl PeerStoreInner {
+	fn is_banned(&self, peer_id: &PeerId) -> bool {
+		self.reputations.get(peer_id).cloned().unwrap_or_default().is_banned()
+	}
+
+	fn report_disconnect(&mut self, peer_id: PeerId, reason: DropReason) {
+		self.reputations
+			.entry(peer_id)
+			.or_default()
+			.add_reputation(DISCONNECT_REPUTATION_CHANGE);
+		// FIXME: in the original `PeerSet`, `DropReason` was used to remove peers
+		// connection attempts to which were rejected. As connection management logic was moved
+		// to `ProtocolController`, I don't think we need this in `PeerStore`.
+	}
+
+	fn outgoing_candidates(&self, count: usize, ignored: HashSet<&PeerId>) -> Vec<PeerId> {
+		let mut candidates = self
+			.reputations
+			.iter()
+			.filter_map(|(peer_id, reputation)| {
+				if !reputation.is_banned() && !ignored.contains(peer_id) {
+					Some((*peer_id, *reputation))
+				} else {
+					None
+				}
+			})
+			.collect::<Vec<_>>();
+		candidates.partial_sort(count, |(_, rep1), (_, rep2)| rep1.cmp(rep2));
+		candidates.iter().take(count).map(|(peer_id, _)| *peer_id).collect()
+
+		// FIXME: depending on usage patterns, it might be more efficient to always maintain peers
+		// sorted.
```

Comment:
I don't think keeping the values in vector and is the right way to go though. If `BinaryHeap` doesn't work, then what about `BTreeMap`? It seems to offer some nice features

Reply:
Update: bimap is not an option, because it wouldn't allow different peers to have equal reputation values. We need something like bimultimap :see_no_evil:

I'm going to postpone this part until we have peering working using new `ProtocolController` and `PeerStore`.

Reply:
For reference: here is the issue for this https://github.com/paritytech/substrate/issues/13903.